doc/dev/osd_internals/manifest.rst

```diff
@@ -287,10 +416,14 @@ Operations:
   TODO: maybe_handle_manifest_detail appears to flush the data prior to
         promoting it, seems unnecessary?
 
+  --> Without this, the dirty never be flush until tier_flush is called explicitly.
```

Comment:
I mean, I don't understand why you'd bother.  Promoting the clean regions would result in all data regions present in the head, right?  I might actually be misunderstanding the value of maintaining dirty regions at all -- a dirty region means that the actual object contents from a librados user point or view does not match the manifest target for that region.  Flushing would actually create a new object with a new region.  So why keep the pointer to the old region at all?

Reply:
This is not related to promoting or something like that. This is for flushing the manifest object to make the manifest object clean. This needs to be removed. But, without this, when and where flush happens for now?

Let's think about the case as above.

1. Object A contains chunk aa ( offset: 10, contents: cccc, state: clean ).
2. Write object A at offset 10.
3. Object A contains chunk aa ( offset: 10, cotents: dddd, state: dirty ).

After object A is overwritten, chunk aa is still dirty and its content is still in the base tier.
Because we don't implement a flush that evicts or flush its contents periodically yet (like agent thread), we added this flush by force. This should be remove after agent thread for manifest object is enabled.

Reply:
Let me ask a simpler question then: consider the object

HEAD: [1~10 manifest: aaa]

If I then perform a write:

write 1~10

the resulting object will be

HEAD: [1~10 manifest: aaa dirty]

with an actual object payload at 1~10 stored in the base pool.  Why don't we just get rid of the manifest mapping for 1~10 entirely and drop the reference?  Why do we care about aaa if it has contents that no longer match the region?

Reply:
Ah, The initial thought behind of this is that we don't know fingerprint oid before flushing the chunk. So, before promoting the chunk, we need to figure out the fingerprint oid. To do that, a flush is needed. 
The reason why we just do not drop the reference is that 1) to give a chance to minimize fingerprint calculation via delaying it until flush happens. 2) moreover, if we perform a write with overlapped range, for example 5~15 write in your case, we need a read at least. It probably trigger read-modify-write whenever the kind of such write comes. Therefore, I thought deferred work is the right way. 
Do you think dropping the reference immediately is better ?

Reply:
If we are going to promote an object with a currently dirty region, I think the current implementation is:

flush (prior to promote):
1) compute the fingerprint for the region
2) write that region to an object with the fingerprint as a name
3) update the base object by setting the manifest to point to that object with state clean
4) decrement the refcount on the old target

promote: noop for this extent because the extent is not missing

Alternately you could:

1) mark the extent as not a manifest extent

If we really want to have the fingerprint precomputed, we could then 
1.1) compute the fingerprint without writing out an actual object

I don't think I understand your 5~15 example.  In either approach, after a promote, all of the data is in the base object.  If the object has:

1~10 present clean
10~10 present clean

and we perform a write spanning them 5~10, I think you'd just update the in place data and remove both extents from the manifest.  In the event that you have

1~10 present clean
10~10 missing

I think you always have to first load the 10~10 portion first, right?

I'm not so much arguing that we should drop the reference, I just don't yet understand what the purpose is.  Is your argument that promote should implicitly ensure that all extents are dedup'd?  For dirty extents outside of promote, I don't see how retaining the old fingerprint helps at all -- it's clearly not correct for that object anymore, and doesn't help with future operations as far as I can tell.  The only information it appears to carry is that we, in the past, dedup'd that extent.  Is that right?

Reply:
1. current implementation is:
flush (prior to promote):
a. compute the fingerprint for the region
_b. decrement the refcount on the old target_
c. write that region to an object with the fingerprint as a name
d. update the base object by setting the manifest to point to that object with state clean _and set new fingerprint oid (old oid is replcaed with new fingerprint oid)

2. Regarding 5 ~ 15 example, 
I mean, I'd prefer to use deferred work for fingerprint calculation unless new fingerprint oid is necessary, For example,
There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
Then, we write the region of 5 ~ 15.
HEAD:[1-10 manifest:ddd, dirty, size is 20]
Then, we write the region of 6 ~ 15.
HEAD:[1-10 manifest:ccc, dirty, size is 20]
Then, we write the region of 7 ~ 15.
HEAD:[1-10 manifest:aaa, dirty, size is 20]
If we want to precompute here, need a read 1~10 and a computation always.


3. I don't understand your case. 
You've explained 

> In either approach, after a promote, all of the data is in the base object. If the object has:
>1 ~ 10 present clean
>10 ~ 10 present clean

Yes

> and we perform a write spanning them 5 ~ 10, I think you'd just update the in place data

Yes, so
1 ~ 10 present dirty
1 ~ 10 present clean

> and remove both extents from the manifest. In the event that you have
>1 ~ 10 present clean
>10 ~ 10 missing

Why a remove happens to both extents? due to flush? or delete op?

Reply:
>     1. current implementation is:
>        flush (prior to promote):
>        a. compute the fingerprint for the region
>        _b. decrement the refcount on the old target_
>        c. write that region to an object with the fingerprint as a name
>        d. update the base object by setting the manifest to point to that object with state clean _and set new fingerprint oid (old oid is replcaed with new fingerprint oid)

So at this point promote is then a noop for that extent, right?  It seems to me that dedup'ing an extent the user has indicated that they want present in the base tier wastes latency, IO, and space.  What if the reason for the promote is that the object is about to see overwrites?  Wouldn't that make the dedup operation a waste?

> 
>     2. Regarding 5 ~ 15 example,
>        I mean, I'd prefer to use deferred work for fingerprint calculation unless new fingerprint oid is necessary, For example,
>        There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
>        Then, we write the region of 5 ~ 15.
>        HEAD:[1-10 manifest:ddd, dirty, size is 20]
>        Then, we write the region of 6 ~ 15.
>        HEAD:[1-10 manifest:ccc, dirty, size is 20]
>        Then, we write the region of 7 ~ 15.
>        HEAD:[1-10 manifest:aaa, dirty, size is 20]
>        If we want to precompute here, need a read 1~10 and a computation always.

Ok, I think this example is the core of what I'm not understanding.  For the 6~15 write,
1) How do you recompute the fingerprint for 1 ~ 10 without looking at 1 ~ 5 (offsets 1 - 6)?
2) If you update the manifest to ccc, don't you also have to write out a new object with name ccc and decrement the ref on ddd?  If so, wouldn't the object be clean?

> 
>     3. I don't understand your case.
>        You've explained
> 
> 
> > In either approach, after a promote, all of the data is in the base object. If the object has:
> > 1 ~ 10 present clean
> > 10 ~ 10 present clean
> 
> Yes
> 
> > and we perform a write spanning them 5 ~ 10, I think you'd just update the in place data
> 
> Yes, so
> 1 ~ 10 present dirty
> 1 ~ 10 present clean

I meant that 5 ~ 10 spans offsets  5 - 15.

> 
> > and remove both extents from the manifest. In the event that you have
> > 1 ~ 10 present clean
> > 10 ~ 10 missing
> 
> Why a remove happens to both extents? due to flush? or delete op?

This was a separate example.

Reply:
>So at this point promote is then a noop for that extent, right? It seems to me that dedup'ing >an extent the user has indicated that they want present in the base tier wastes latency, IO, >and space. What if the reason for the promote is that the object is about to see overwrites? >Wouldn't that make the dedup operation a waste since we'd then have to mark the extent dirty.

Yes, promote is a noop. 
If the purpose of promote is just for overwrites, we can re-use existing op like CACHE_PIN to
avoid dedup operations. 

>Ok, I think this example is the core of what I'm not understanding. For the 6~15 write,

>How do you recompute the fingerprint for 1 ~ 10 without looking at 1 ~ 5 (offsets 1 - 6)?

Read entire range (1 ~ 10) and then compute the fingperint.
I meant, we need to read whole chunk and then a computation in case of the partial write if pre-computation is used.

>If you update the manifest to ccc, don't you also have to write out a new object with name >ccc and decrement the ref on ddd? If so, wouldn't the object be clean?

Yes, right.  HEAD:[1-10 manifest:ccc, clean, size is 20]. My intention is that we need a read -- computation operation every partial write if pre-computation is used.  
In short,

-- Pre-computation
There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
Then, we write the region of 5 ~ 15.
And, Read 1-10 and computation.
HEAD:[1-10 manifest:bbb, clean, size is 20]
Then, we write the region of 6 ~ 15.
And, Read 1-10 and computation.
HEAD:[1-10 manifest:ccc, clean, size is 20]
Then, we write the region of 7 ~ 15.
And, Read 1-10 and computation
HEAD:[1-10 manifest:ddd, clean, size is 20]

-- Current implementation
There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
Then, we write the region of 5 ~ 15.
HEAD:[1-10 manifest: aaa, dirty, size is 20]
Then, we write the region of 6 ~ 15.
HEAD:[1-10 manifest: aaa, dirty, size is 20]
Then, we write the region of 7 ~ 15.
If flush happens or new fingerprint oid is needed,
Read 1-10 and computation
HEAD:[1-10 manifest:ddd, dirty, size is 20]

As a result, we can save two reads and two computations.

> I meant that 5 ~ 10 spans offsets 5 - 15.

In that case, should we return ERANGE and unset-manifest first.
Current implementation does not allow overlapped chunk range.
For example, there are 5 ~ 10 and  10 ~ 15. 
5 ~ 10 can not become 5 ~ 15.

>1 ~ 10 present clean
>10 ~ 10 missing
>I think you always have to first load the 10 ~ 10 portion first, right?

Right

>Is your argument that promote should implicitly ensure that all extents are dedup'd? For dirty extents outside of promote, I don't see how retaining the old fingerprint helps at all -- it's clearly not correct for that object anymore, and doesn't help with future operations as far as I can tell. The only information it appears to carry is that we, in the past, dedup'd that extent. Is that right?

I explained above, this is because the design regarding deferred computation. 
By using this scheme, we can save reads and computations in the case of partial writes,
This can be optimized as follows.

--As is
1. invoke promote() <- promote will start after flush is finished
2. flush
3. do promote()

--To be
1. invoke promote()
2. compute a fingerprint without flush
3. do promote() based on the fingerprint from 2

Reply:
> > So at this point promote is then a noop for that extent, right? It seems to me that dedup'ing >an extent the user has indicated that they want present in the base tier wastes latency, IO, >and space. What if the reason for the promote is that the object is about to see overwrites? >Wouldn't that make the dedup operation a waste since we'd then have to mark the extent dirty.
> 
> Yes, promote is a noop.
> If the purpose of promote is just for overwrites, we can re-use existing op like CACHE_PIN to
> avoid dedup operations.

How would that work?  It seems to me that I'd have to promote first and then use CACHE_PIN in that case, which would still force an unnecessary sequence of dedup operations.  I don't see why you'd default to always deduping when the user specifically asked for the data to be resident in the base pool.  After all, they could use promote and then set_chunk on extents which happen not to be dedup'd, right?

> 
> > Ok, I think this example is the core of what I'm not understanding. For the 6~15 write,
> 
> > How do you recompute the fingerprint for 1 ~ 10 without looking at 1 ~ 5 (offsets 1 - 6)?
> 
> Read entire range (1 ~ 10) and then compute the fingperint.
> I meant, we need to read whole chunk and then a computation in case of the partial write if pre-computation is used.
> 
> > If you update the manifest to ccc, don't you also have to write out a new object with name >ccc and decrement the ref on ddd? If so, wouldn't the object be clean?
> 
> Yes, right. HEAD:[1-10 manifest:ccc, clean, size is 20]. My intention is that we need a read -- computation operation every partial write if pre-computation is used.
> In short,
> 
> -- Pre-computation
> There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
> Then, we write the region of 5 ~ 15.
> And, Read 1-10 and computation.
> HEAD:[1-10 manifest:bbb, clean, size is 20]
> Then, we write the region of 6 ~ 15.
> And, Read 1-10 and computation.
> HEAD:[1-10 manifest:ccc, clean, size is 20]
> Then, we write the region of 7 ~ 15.
> And, Read 1-10 and computation
> HEAD:[1-10 manifest:ddd, clean, size is 20]
> 
> -- Current implementation
> There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
> Then, we write the region of 5 ~ 15.
> HEAD:[1-10 manifest: aaa, dirty, size is 20]
> Then, we write the region of 6 ~ 15.
> HEAD:[1-10 manifest: aaa, dirty, size is 20]
> Then, we write the region of 7 ~ 15.
> If flush happens or new fingerprint oid is needed,
> Read 1-10 and computation
> HEAD:[1-10 manifest:ddd, dirty, size is 20]

wouldn't this be

HEAD:[1-10 manifest:ddd, clean, size is 20]

> 
> As a result, we can save two reads and two computations.

Sure, I see why we don't want to perform a dedup operation betwen every pair.  What if we instead:
There is a HEAD: [1-10 manifest: aaa, clean, size is 20]
Then, we write the region of 5 ~ 15.
HEAD:[size is 20] (1-10 is nochunk)
Then, we write the region of 6 ~ 15.
HEAD:[size is 20] (1-10 is nochunk)
Then, we write the region of 7 ~ 15.
If the tiering agent wants to dedup 1-10 because it is now cold, it can use a read and set-chunk to:
Read 1-10 and computation
HEAD:[1-10 manifest:ddd, clean, size is 20]

This way, we shorten the lifetime of the aaa dedup target object freeing space earlier.  Can't you can always do this instead of marking an extent dirty?

> 
> > I meant that 5 ~ 10 spans offsets 5 - 15.
> 
> Yes, so it will become
> 1 ~ 10 present dirty
> 10 ~ 10 present dirty
> Your question is that we need to drop the old reference at this point?
> If many partial write comes continuously, that cause a serious overhead.
> How is making two modes there: buffered and direct when set_chunk is set.
> Direct always update the fingerprint whenever writes come.
> On the other hand, buffered delay those tasks until some event is triggered.
> 

See my above example, I don't see why dropping the refcount causes extra work.

> > 1 ~ 10 present clean
> > 10 ~ 10 missing
> > I think you always have to first load the 10 ~ 10 portion first, right?
> 
> Right
> 
> > Is your argument that promote should implicitly ensure that all extents are dedup'd? For dirty extents outside of promote, I don't see how retaining the old fingerprint helps at all -- it's clearly not correct for that object anymore, and doesn't help with future operations as far as I can tell. The only information it appears to carry is that we, in the past, dedup'd that extent. Is that right?
> 
> I explained above, this is because the design regarding deferred computation.
> By using this scheme, we can save reads and computations in the case of partial writes,
> This can be optimized as follows.
> 
> --As is
> 
>     1. invoke promote() <- promote will start after flush is finished
> 
>     2. flush
> 
>     3. do promote()
> 
> 
> --To be
> 
>     1. invoke promote()
> 
>     2. compute a fingerprint without flush
> 
>     3. do promote() based on the fingerprint from 2

Reply:
>How would that work? It seems to me that I'd have to promote first and then use ?>CACHE_PIN in that case, which would still force an unnecessary sequence of dedup >operations. I don't see why you'd default to always deduping when the user specifically >asked for the data to be resident in the base pool. After all, they could use promote and >then set_chunk on extents which happen not to be dedup'd, right?

I mean, using TIER_DO_NOT_DEDUP_PIN (like CACHE_PIN), we can indicate whether the target object needs to be deduped or not.
As I described above, there are two users case.  Based on these,
The first case is:
1. Create Object A and B
2. Setchunk A to B
3. Write A
4. TIER_DO_NOT_DEDUP_PIN
5. Flush does not occur

The second case is:
1. Create Object A and B
2. Setchunk A to B
3. Promote A
4. Write A
5. TIER_DO_NOT_DEDUP_PIN
6. Flush does not occur 


>This way, we shorten the lifetime of the aaa dedup target object freeing space earlier. I >don't see how remembering that, in the past, the dedup target happened to be aaa helps in >recomputing and deduping 1-10 after the 7~15 write.

Right, in that case, freeing space earlier is better than releasing the space when flush happens. I will consider it when creating PRs with regard to clean-up.

Reply:
In your second case, promote still unnecessarily dedups A.

When you consider the free space problem, also consider whether there are *any* situations where DIRTY would be preferable to simply clearing the chunk (NO CHUNK).  If there are no such cases, we can completely eliminate it as a case.  My goal with this conversation is to ensure that the implementation is as simple as possible, and removing DIRTY as a case throughout the code would be a big simplification if there are no downsides.

Reply:
Yeah, but
If the case is,

Create Object A and B
Setchunk A to B
TIER_DO_NOT_DEDUP_PIN
Promote A
Write A
Flush does not occur

It probably promotes chunks to Object A because chunks in Object A are missing.
But, it does not dedup due to TIER_DO_NOT_DEDUP_PIN.

>When you consider the free space problem, also consider whether there are any situations ?>where DIRTY would be preferable to simply clearing the chunk (NO CHUNK). If there are no >such cases, we can completely eliminate it as a case. My goal with this conversation is to >ensure that the implementation is as simple as possible, and removing DIRTY as a case >throughout the code would be a big simplification if there are no downsides.
Ok.

Reply:
I agree that that behaviour could be equivalent, but I consider it highly unintuitive.  If you want it to work that way, you'll have to carefully document this interface to ensure that developers debugging and using this interface won't make the assumptions I did.